query-engine/connectors/mongodb-query-connector/src/filter.rs

```diff
@@ -188,42 +188,53 @@ impl MongoFilterVisitor {
             // Todo: The nested list unpack looks like a bug somewhere.
             //       Likely join code mistakenly repacks a list into a list of PrismaValue somewhere in the core.
             ScalarCondition::In(vals) => match vals {
-                ConditionListValue::List(vals) => match vals.split_first() {
-                    // List is list of lists, we need to flatten.
-                    Some((PrismaValue::List(_), _)) => {
-                        let mut bson_values = Vec::with_capacity(vals.len());
-
-                        for pv in vals {
-                            if let PrismaValue::List(inner) = pv {
-                                bson_values.extend(
-                                    inner
-                                        .into_iter()
-                                        .map(|val| self.coerce_to_bson_for_filter(field, val))
-                                        .collect::<crate::Result<Vec<_>>>()?,
-                                )
-                            }
+                ConditionListValue::List(values) => {
+                    let mut equalities = Vec::with_capacity(values.len());
+
+                    for value in values {
+                        // List is list of lists, we need to flatten.
+                        // This flattening behaviour does not affect user queries because Prisma does
+                        // not support storing arrays as values inside a field. It is possible to have
+                        // a 1-dimensional array field, but not 2 dimensional. Thus, we never have
+                        // user queries which have arrays in the argument of `in` operator. If we
+                        // encounter such case, then this query was produced internally and we can
+                        // safely flatten it.
+                        if let PrismaValue::List(list) = value {
+                            equalities.extend(
+                                list.into_iter()
+                                    .map(|value| {
+                                        let value = self.coerce_to_bson_for_filter(field, value)?;
+                                        Ok(doc! { "$eq": [&field_name, value] })
+                                    })
+                                    .collect::<crate::Result<Vec<_>>>()?,
+                            );
+                        } else {
+                            let value = self.coerce_to_bson_for_filter(field, value)?;
+                            equalities.push(doc! { "$eq": [&field_name, value] })
                         }
-
-                        doc! { "$in": [&field_name, bson_values] }
-                    }
-                    _ => {
-                        doc! { "$in": [&field_name, self.coerce_to_bson_for_filter(field, PrismaValue::List(vals))?] }
                     }
-                },
+
+                    doc! { "$or": equalities }
```

Comment:
Could you perhaps add a small comment that retains the info on why we switched to `$or`?